### ISSUE TITLE
Support Python 3.11

### ISSUE THREAD

**Issue by Innixma** [autogluon/autogluon#2687]
Add support for Python 3.11 in AutoGluon.

Tentatively planned for AutoGluon v0.8 release in ~June 2023. Unknown if there will be blockers at this time.

- [x] Pre-req: Add Python 3.10 support (https://github.com/autogluon/autogluon/issues/2686)
- [x] https://github.com/autogluon/autogluon/pull/2739
- [x] https://github.com/catboost/catboost/issues/2213
  - [x] Note: Latest CatBoost seems to break on MacOS when Python != 3.11, need to investigate. 
- [x] ray 3.11 support
- [x] torchtext 3.11 support
- [x] torchvision 3.11 support

**Comment by gradientsky** [comment]
Must be updated
`torch>=1.13.1`

Not supported:
`ray`
`catboost`
`torchtext`
`torchvision`

**Comment by Innixma** [comment]
CatBoost plans to support Python 3.11 soon, likely in the next few months: https://github.com/catboost/catboost/issues/2213

**Comment by ddelange** [comment]
torch is done: https://github.com/autogluon/autogluon/pull/2739

are there any specific reasons, apart from its dependencies, to enforce the upper python version in the autogluon packages?

if one were to try to install autogluon on cp311, and autogluon did not have this upper cap, pip would simply fail on one of the dependencies not supporting cp311. should be fine right?

**Comment by Innixma** [comment]
> torch is done: #2739

Thanks for notifying!

> are there any specific reasons, apart from its dependencies, to enforce the upper python version in the autogluon packages?
> 
> if one were to try to install autogluon on cp311, and autogluon did not have this upper cap, pip would simply fail on one of the dependencies not supporting cp311. should be fine right?

It is better to avoid confusion from users if we don't actually support 3.11. To have it crash on one of the inner dependencies install is equivalent to AutoGluon succeeding install and then crashing on import due to incompatible dependencies. In both cases it is misleading the user into thinking it is somehow their fault for the error. I think it is good we maintain an upper python version cap, and expand the cap once we support a new version.

**Comment by ddelange** [comment]
`ray~=2.3` [ships](https://pypi.org/project/ray/2.3.0/#files) cp311
`torchtext~=0.15` [ships](https://download.pytorch.org/whl/torchtext/) cp311
`torchvision~=0.15` [ships](https://download.pytorch.org/whl/torchvision/) cp311

the latter two require `torch==2.0.0`:
```
Metadata-Version: 2.1
Name: torchvision
Version: 0.15.1
Summary: image and video datasets and models for torch deep learning
Home-page: https://github.com/pytorch/vision
Author: PyTorch Core Team
Author-email: soumith@pytorch.org
License: BSD
Requires-Python: >=3.8
License-File: LICENSE
Requires-Dist: numpy
Requires-Dist: requests
Requires-Dist: torch (==2.0.0)
Requires-Dist: pillow (!=8.3.*,>=5.3.0)
Provides-Extra: scipy
Requires-Dist: scipy ; extra == 'scipy'
```

**Comment by ddelange** [comment]
this also implies a CUDA verion bump to `cu117` or `cu118`. there's also CUDA 12 support in the pipeline: https://github.com/pytorch/pytorch/issues/91122

**Comment by ddelange** [comment]
cc @sxjscience

**Comment by Innixma** [comment]
CatBoost 1.2 released today, adding Python 3.11 support: https://github.com/catboost/catboost/releases/tag/v1.2

**Comment by gradientsky** [comment]
@ddelange `torch==2.0.0` would be a milestone on it's own

**Comment by ddelange** [comment]
there's quite a chance that pt 2.0 will 'just work'. there's now `zero_grad(set_to_none=True)` (default used to be False), for the rest the changes [look](https://github.com/pytorch/pytorch/releases/tag/v2.0.0) all good to me

**Comment by ddelange** [comment]
torchmetrics pt 2.0 support is merged but unreleased: https://github.com/Lightning-AI/torchmetrics/commit/95599c9a7bd5e6c8f0a8039562dbbc832b434031

torch lightning supports pt 2.0 as of v2.0.0: https://github.com/Lightning-AI/lightning/commit/ef2a6088ff5bc515be27e24f5f6c8b16b96d3bb8

**Comment by Innixma** [comment]
Added in #3190! Thanks to @ddelange, @yinweisu, and everyone else who helped with the Python 3.11 support!

**Comment by ddelange** [comment]
one leftover is to support `optimize_for_inference` on py 3.11:

tensorrt 8.6.1 adds support for py 3.11, but was causing some issues in autogluon CI which were not investigated yet

**Comment by Innixma** [comment]
Re-opening to track @ddelange's comment on tensorrt, thanks for the heads up!

**Comment by ddelange** [comment]
fwiw, tensorrt 9.1 was just [released](https://github.com/NVIDIA/TensorRT/releases/tag/v9.1.0), pypi should follow soon

**Comment by ddelange** [comment]
another leftover is to support the ray 2.7.0 release:

it contains some show stopper bugfixes for ray serve, which are specific to python 3.11 ref https://github.com/kserve/kserve/pull/3075#issuecomment-1677630840

**Comment by ddelange** [comment]
> one leftover is to support `optimize_for_inference` on py 3.11:
> 
> tensorrt 8.6.1 adds support for py 3.11, but was causing some issues in autogluon CI which were not investigated yet

update from @AnirudhDagar: https://github.com/autogluon/autogluon/pull/3190#discussion_r1415639872

tensorrt is blocked by cuda 12.1 / torch 2.1 ref #3663

**Comment by drewbitt** [comment]
#3982 would unblock this.

**Comment by Innixma** [comment]
I believe this is now resolved thanks to the torch 2.1 and ray 2.10 upgrades, please re-open in case we identify any issues.

Thanks everyone for the assistance, especially @ddelange!
